File: gui_v2/src/types/domain/placement.ts
/**
 * Placement declarations — where work may run, and for how long.
 *
 * The TypeScript twin of `agent/placement.py`. A skill declares what it
 * `requires`; a pod advertises what it `capabilities`. The scheduler matches
 * one against the other, so the two lists must use the same words — a
 * requirement no pod can advertise is a turn that queues forever.
 *
 * Keep CAPABILITIES in step with `agent/placement.py`.
 */

export type Residency = 'local' | 'cloud' | 'any';
export type Lifetime = 'turn' | 'conversation' | 'long_running';

export const DEFAULT_RESIDENCY: Residency = 'any';
export const DEFAULT_LIFETIME: Lifetime = 'conversation';

export interface Placement {
  residency: Residency;
  lifetime: Lifetime;
  requires: string[];
}

export const DEFAULT_PLACEMENT: Placement = {
  residency: DEFAULT_RESIDENCY,
  lifetime: DEFAULT_LIFETIME,
  requires: [],
};

/** What a pod can advertise and a skill can require. */
export const CAPABILITIES = [
  'browser_local',
  'gpu',
  'cn_llm',
  'intl_llm',
  'long_running',
  'file_storage',
] as const;

export type Capability = (typeof CAPABILITIES)[number];

export const CAPABILITY_LABELS: Record<string, string> = {
  browser_local: 'Local browser (RPA / browser-use)',
  gpu: 'GPU inference',
  cn_llm: 'CN-region LLM routing',
  intl_llm: 'International LLM routing',
  long_running: 'Hosts work that never checkpoints',
  file_storage: 'Durable local scratch space',
};

/**
 * Isolation as a capability, never as a binding: a pod carrying
 * `dedicated:<agent-id>` is the only one that can claim that agent's work, but
 * it is still a scheduling decision — if the pod is down the work waits rather
 * than being stranded on a dead machine.
 */
export const DEDICATED_PREFIX = 'dedicated:';

export const dedicatedCapability = (agentId: string): string =>
  `${DEDICATED_PREFIX}${String(agentId || '').trim()}`;

export const isDedicatedCapability = (value: string): boolean =>
  String(value || '').startsWith(DEDICATED_PREFIX);

export const RESIDENCY_OPTIONS: { value: Residency; label: string; help: string }[] = [
  {
    value: 'any',
    label: 'Any',
    help: 'The scheduler places this wherever there is room. The default.',
  },
  {
    value: 'local',
    label: 'Local only',
    help: 'Runs on this machine. Choose this when the work needs something only this desktop has.',
  },
  {
    value: 'cloud',
    label: 'Cloud only',
    help: 'Runs on a pod. Choose this when the work should keep running with the desktop closed.',
  },
];

export const LIFETIME_OPTIONS: { value: Lifetime; label: string; help: string }[] = [
  {
    value: 'turn',
    label: 'Turn — one request, then done',
    help: 'Fine on a cold pod: the ~80s startup does not matter for work nobody is waiting on.',
  },
  {
    value: 'conversation',
    label: 'Conversation — many turns over time',
    help: 'Needs a warm pod. This is chat: somebody is waiting, so a cold start is felt.',
  },
  {
    value: 'long_running',
    label: 'Long running — holds a pod for its whole life',
    help: 'For work that cannot be checkpointed mid-run, such as a crawler holding a browser. It occupies a pod until it finishes.',
  },
];

/** Normalize anything stored on a skill/task into a complete Placement. */
export function readPlacement(source: Record<string, any> | null | undefined): Placement {
  const src = source || {};
  const residency = String(src.residency || '').toLowerCase();
  const lifetime = String(src.lifetime || '').toLowerCase();
  const requires = Array.isArray(src.requires)
    ? src.requires.map((r: any) => String(r || '').trim()).filter(Boolean)
    : [];
  return {
    residency: (['local', 'cloud', 'any'].includes(residency)
      ? residency
      : DEFAULT_RESIDENCY) as Residency,
    lifetime: (['turn', 'conversation', 'long_running'].includes(lifetime)
      ? lifetime
      : DEFAULT_LIFETIME) as Lifetime,
    requires: Array.from(new Set(requires)),
  };
}

/** True when a placement says nothing a scheduler could act on. */
export function isDefaultPlacement(p: Placement): boolean {
  return (
    p.residency === DEFAULT_RESIDENCY &&
    p.lifetime === DEFAULT_LIFETIME &&
    p.requires.length === 0
  );
}

File: agent/placement.py
"""Placement declarations — one reader, one vocabulary.

``residency`` / ``lifetime`` / ``requires[]`` exist on ``EC_Skill`` and
``ManagedTask`` (Path 1.5 Phase 0.3) and, until now, nothing read them. Under
unified execution the scheduler does: ``requires[]`` is matched against a pod's
``capabilities[]`` as a WHERE clause, and ``lifetime`` decides whether work can
go to a cold pod at all.

Everything that reads or writes a placement goes through here, for one reason:
these values are persisted **inside the skill's ``config`` JSON**, because
GraphQL ``SkillUpdateInput`` has no columns for them (the same reason
``run_in_cloud`` lives there). That means every load path has to look in two
places and agree on the precedence, and the last time a field like this was
added it silently no-op'd in three separate metadata whitelists. One reader is
the fix.

The capability vocabulary lives here too, and its TypeScript twin is
``gui_v2/src/types/domain/placement.ts``. A skill can only usefully require what
some pod can advertise, so the two lists have to say the same words — a
requirement nothing can satisfy is a turn that queues forever.
"""
from __future__ import annotations

from typing import Any, Dict, Iterable, List, Optional

from utils.logger_helper import logger_helper as logger

RESIDENCY_VALUES = ("local", "cloud", "any")
LIFETIME_VALUES = ("turn", "conversation", "long_running")

DEFAULT_RESIDENCY = "any"
DEFAULT_LIFETIME = "conversation"

# Capabilities a pod can advertise and a skill can require. Keep in step with
# gui_v2/src/types/domain/placement.ts.
CAPABILITIES = (
    "browser_local",     # a real browser on the machine (RPA, browser-use)
    "gpu",               # GPU-backed inference
    "cn_llm",            # CN-region LLM routing
    "intl_llm",          # international LLM routing
    "long_running",      # will host work that never checkpoints
    "file_storage",      # durable local scratch space
)

# Isolation is expressed as a capability, never as a binding: a pod that carries
# dedicated:<agent-id> is the only one that can claim work requiring it, but the
# relationship is still scheduling, so nothing strands when that pod dies —
# the work waits rather than disappearing.
DEDICATED_PREFIX = "dedicated:"

MAX_REQUIRES = 20


def dedicated_capability(agent_id: Any) -> str:
    """The capability that expresses "only this agent's work"."""
    agent = str(agent_id or "").strip()
    if not agent:
        raise ValueError("dedicated capability needs an agent id")
    return f"{DEDICATED_PREFIX}{agent}"


def is_dedicated_capability(value: Any) -> bool:
    return str(value or "").startswith(DEDICATED_PREFIX)


def normalize_residency(value: Any, default: str = DEFAULT_RESIDENCY) -> str:
    text = str(value or "").strip().lower()
    if text in RESIDENCY_VALUES:
        return text
    if text:
        logger.warning(f"[placement] unknown residency {value!r}; using {default!r}")
    return default


def normalize_lifetime(value: Any, default: str = DEFAULT_LIFETIME) -> str:
    text = str(value or "").strip().lower()
    if text in LIFETIME_VALUES:
        return text
    if text:
        logger.warning(f"[placement] unknown lifetime {value!r}; using {default!r}")
    return default


def normalize_requires(value: Any) -> List[str]:
    """A clean capability list: strings, de-duplicated, order kept, bounded.

    Unknown names are NOT dropped — ``dedicated:<agent-id>`` is a legitimate
    requirement that no static list can contain, and a pod may advertise a
    capability this build has never heard of. Dropping them here would silently
    widen placement, which is the dangerous direction.
    """
    if isinstance(value, str):
        items: Iterable[Any] = [v for v in value.replace(";", ",").split(",")]
    elif isinstance(value, (list, tuple, set)):
        items = value
    else:
        return []

    seen: set = set()
    out: List[str] = []
    for item in items:
        text = str(item or "").strip()
        if not text or text in seen:
            continue
        seen.add(text)
        out.append(text)
        if len(out) >= MAX_REQUIRES:
            break
    return out


def read_placement(*sources: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    """Placement from the first source that names each field.

    Sources are consulted in order, which is how the existing cloud flags
    behave: a file's top-level keys win over the ``config`` dict they were also
    written into, so an edit on disk is not shadowed by a stale DB copy.
    """
    residency: Any = None
    lifetime: Any = None
    requires: Any = None

    for source in sources:
        if not isinstance(source, dict):
            continue
        if residency is None and source.get("residency") is not None:
            residency = source.get("residency")
        if lifetime is None and source.get("lifetime") is not None:
            lifetime = source.get("lifetime")
        if requires is None and source.get("requires") is not None:
            requires = source.get("requires")

    return {
        "residency": normalize_residency(residency),
        "lifetime": normalize_lifetime(lifetime),
        "requires": normalize_requires(requires),
    }


def apply_placement(target: Any, *sources: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    """Read a placement onto ``target`` (an EC_Skill / ManagedTask). Returns it."""
    placement = read_placement(*sources)
    try:
        target.residency = placement["residency"]
        target.lifetime = placement["lifetime"]
        target.requires = list(placement["requires"])
    except Exception as exc:                       # a frozen or odd model
        logger.warning(f"[placement] could not apply placement to {type(target).__name__}: {exc}")
    return placement


def placement_of(obj: Any) -> Dict[str, Any]:
    """The placement an object currently declares, defaults filled in."""
    return {
        "residency": normalize_residency(getattr(obj, "residency", None)),
        "lifetime": normalize_lifetime(getattr(obj, "lifetime", None)),
        "requires": normalize_requires(getattr(obj, "requires", None)),
    }


def is_default_placement(placement: Dict[str, Any]) -> bool:
    """True when a placement says nothing a scheduler could act on."""
    return (
        placement.get("residency") == DEFAULT_RESIDENCY
        and placement.get("lifetime") == DEFAULT_LIFETIME
        and not placement.get("requires")
    )
